homeassistant/components/sensor/yr.py

```diff
@@ -92,10 +92,8 @@ def async_setup_platform(hass, config, async_add_devices, discovery_info=None):
 
     weather = YrData(hass, coordinates, forecast, dev)
     # Update weather on the hour, spread seconds
-    async_track_utc_time_change(
-        hass, weather.async_update, minute=randrange(1, 10),
-        second=randrange(0, 59))
-    yield from weather.async_update()
+    async_track_utc_time_change(hass, weather.async_update, second=0)
```

Comment:
this is not on the hour. This is called every minute.

Reply:
Read the code, the comment is stale.

Reply:
So this is very inefficient. Instead of calling update every minute, we should just keep track when we want to fetch data, then when data is fetched, we should call the update method.

I think that it should be something like this:

```python
def async_setup_platform(…):
    async_call_later(hass, 5, weather.update_data)

class YrData(object):
    @asyncio.coroutine
    def update_data(self):
        try:
            yield from fetching_data()
            yield from updating_devices()
            async_call_later(hass, random(X), self.update_data)
        except …:
            async_call_later(hass, random(15), self.update_data)
```

Reply:
Wait, it seems like you're doing this already. So this line can be removed

Reply:
We want to call `updating_devices()` more often and more regular than `fetching_data()`, but every minute is probably overkill.

When we fetch the data we get the forecast for several hours:
Eg: 10:00, 11:00, 12:00, 13:00, 14:00,
So at 10:31 we want to show the 11:00 forecast as that is closest in time.

We still want to fetch new data each hour, to make sure we have newest and best forecast.

Reply:
Ah, I see. In a perfect world we would set a timer for exactly that time to push an update.